nexusparser/tools/yaml2nxdl/yaml2nxdl.py

```diff
@@ -1,88 +1,114 @@
 from yaml2nxdl_utils import nx_type_keys, nx_attr_idnt
-from yaml2nxdl_read_user_yml_appdef import read_user_appdef
+from yaml2nxdl_read_yml_appdef import read_application_definition
 from yaml2nxdl_recursive_build import recursive_build
 
-class yml2nxdl():
-    def __init__(self, fn, *args, **kwargs):
-        #add check if file exists
-        #step1: read the user-specific application definition which was written as a yml file
-        self.fnm = fn
-        self.yml = read_user_appdef(self.fnm)
-        print('Read YAML schema file')
+from typing import Tuple
+import click
 
-    def parse(self):
-        #step2a: create an instantiated NXDL schema XML tree, begin with the header add XML schema/namespaces
-        attr_qname = etree.QName("http://www.w3.org/2001/XMLSchema-instance", "schemaLocation")
-        rt = etree.Element('definition', 
-                   {attr_qname: 'http://definition.nexusformat.org/nxdl/nxdl.xsd' },
-                   nsmap = {None: 'http://definition.nexusformat.org/nxdl/3.1', 
-                            'xsi': 'http://www.w3.org/2001/XMLSchema-instance'})
-        #step2b: user-defined attributes for the root group
-        if 'name' in self.yml.keys():
-            rt.set('name', self.yml['name'])
-            del self.yml['name']
-        else:
-            raise ValueError('ERROR: name: keyword not specified !')
-        pi = etree.ProcessingInstruction("xml-stylesheet", text='type="text/xsl" href="nxdlformat.xsl"')
-        rt.addprevious(pi)
-        #evaluate whether we handle an application definition, a contributed or base class
-        if 'category' in self.yml.keys():
-            if self.yml['category'] == 'application':
-                rt.set('category', 'application')
-                rt.set('extends', 'NXentry')
-            elif self.yml['category'] == 'contributed':
-                rt.set('category', 'contributed')
-                rt.set('extends', 'NXobject')
-            elif self.yml['category'] == 'base':
-                rt.set('category', 'base')
-                rt.set('extends', 'NXobject')
-            else:
-                raise ValueError('Top-level keyword category exists in the yml but one of these: application, contributed, base !')
-            del self.yml['category']
-            rt.set('type', 'group')
-        else:
-            raise ValueError('Top-level keyword category does not exist in the yml !')
-        #step2c: docstring
-        if 'doc' in self.yml.keys():
-            rt.set('doc', self.yml['doc'])
-            del self.yml['doc']
+
+def pretty_print_xml_given_root(root, output_xml):
+    """
+    Print formatted xml file with built-in libraries
+    """
+    xml_string = xml.dom.minidom.parseString(ET.tostring(root)).toprettyxml()
+    xml_string = os.linesep.join([s for s in xml_string.splitlines() if s.strip()])
+    with open(output_xml, "w") as file_out:
+        file_out.write(xml_string)
```

Comment:
Too long names are better than too short one. But just for the "feeling", I would have gone for `pretty_print_xml`. If you later use good names for the root, you could end up with something like `pretty_print_xml(xml_root)` or something.